src/org/sipdroid/sipua/PublishAgent.java

```diff
@@ -0,0 +1,218 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This source code is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this source code; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * Author(s):
+ * Pranav Jain (contact@pranavjain.me)
+ */
+
+package org.sipdroid.sipua;
+
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+
+import org.zoolu.sip.address.NameAddress;
+import org.zoolu.sip.authentication.DigestAuthentication;
+import org.zoolu.sip.header.AuthorizationHeader;
+import org.zoolu.sip.header.ProxyAuthenticateHeader;
+import org.zoolu.sip.header.ProxyAuthorizationHeader;
+import org.zoolu.sip.header.StatusLine;
+import org.zoolu.sip.header.ViaHeader;
+import org.zoolu.sip.header.WwwAuthenticateHeader;
+import org.zoolu.sip.message.Message;
+import org.zoolu.sip.message.MessageFactory;
+import org.zoolu.sip.provider.SipProvider;
+import org.zoolu.sip.provider.SipStack;
+import org.zoolu.sip.transaction.TransactionClient;
+import org.zoolu.sip.transaction.TransactionClientListener;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.logging.Logger;
+
+
+public class PublishAgent implements TransactionClientListener {
+
+	/**
+	 * UserProfile
+	 */
+	protected UserAgentProfile user_profile;
+	/**
+	 * User name.
+	 */
+	String username;
+
+	/**
+	 * User realm.
+	 */
+	String realm;
+	/**
+	 * Nonce for the next authentication.
+	 */
+	String next_nonce;
+	/**
+	 * Qop for the next authentication.
+	 */
+	String qop;
+	/**
+	 * User's passwd.
+	 */
+	String passwd;
+	/**
+	 * Expiration time.
+	 */
+	int expire_time;
+
+	private Logger logger = Logger.getLogger(getClass().getCanonicalName());
+	/**
+	 * SipProvider
+	 */
+	protected SipProvider sip_provider;
+	Context context;
+
+	/**
+	 * Costructs a new MessageAgent.
+	 */
+	public PublishAgent(SipProvider sip_provider, UserAgentProfile user_profile, String username, String realm, String passwd, Context context) {
+		this.sip_provider = sip_provider;
+		this.user_profile = user_profile;
+		this.username = username;
+		this.realm = realm;
+		this.passwd = passwd;
+		this.next_nonce = null;
+		this.qop = null;
+		this.context = context;
+
+	}
+
+	public void publish() {
+		this.expire_time=SipStack.default_expires;
+		this.publish("open", expire_time, " ");
+	}
+
+	public void publish(String status, int expireTime, String note) {

```

Comment:
You still don't validate input. `status` must **only** be one of "open" or "closed".